sigstore/_internal/fulcio/_client.py

```diff
@@ -80,40 +87,61 @@ def root_cert(self) -> Endpoint:
 
 
 class FulcioSigningCert(Endpoint):
-    def post(self, req: CertificateRequest, token: str) -> CertificateResponse:
-        """Get the signing certificate"""
-        resp: requests.Response = self.session.post(
-            url=self.url,
-            data=req.json(),
-            headers={"Authorization": f"Bearer {token}", "Content-Type": "application/json"},
-        )
+    def post(
+        self, req: FulcioCertificateSigningRequest, token: str
+    ) -> FulcioCertificateSigningResponse:
+        """
+        Get the signing certificate.
+
+        Ideally, in the future, this could take an X.509 Certificate Signing
+        Request object instead [^1], but the Fulcio API doesn't currently
+        support this [^2].
+
+        [^1]: https://cryptography.io/en/latest/x509/reference/#x-509-csr-certificate-signing-request-object  # noqa
+        [^2]: https://github.com/sigstore/fulcio/issues/503
+
+        """
+        headers = {
+            "Authorization": f"Bearer {token}",
+            "Content-Type": "application/json",
+            "Accept": "application/pem-certificate-chain",
+        }
+        resp: requests.Response = self.session.post(url=self.url, data=req.data, headers=headers)
         try:
             resp.raise_for_status()
         except requests.HTTPError as http_error:
-            raise FulcioClientError from http_error
+            try:
+                text = json.loads(http_error.response.text)
+                raise FulcioClientError(text["message"]) from http_error
+            except (AttributeError, KeyError):
+                raise FulcioClientError from http_error
+
         sct: str
         try:
             sct = resp.headers["SCT"]
         except IndexError as index_error:
             raise FulcioClientError from index_error
-        pem_blocks = resp.raw.split(PEM_BLOCK_DELIM)
-        if len(pem_blocks) != 3 or not pem_blocks[0]:
+
+        # Cryptography doesn't have chain verification/building built in
+        # https://github.com/pyca/cryptography/issues/2381
+
+        pem_blocks = resp.text.split(CERT_START.decode())
+        if not pem_blocks or pem_blocks[0]:
             raise FulcioClientError(f"Unexpected number of PEM blocks in Fulcio response: {resp}")
-        pem_blocks = pem_blocks[1:]
-        cert_list: List[Certificate] = []
-        for pem_block in pem_blocks:
-            cert: Certificate = load_pem_x509_certificate(pem_block)
-            cert_list.append(cert)
-        return CertificateResponse(cert_list, sct)
+        cert_list: List[Certificate] = [
```

Comment:
Can we keep the check for 2 certificates or can there be an arbitrary number? The Go client only expects two: https://github.com/sigstore/fulcio/blob/main/pkg/api/client.go#L120-L125

Reply:
I don't think that's what's happening there, [`pem.Decode`](https://pkg.go.dev/encoding/pem#Decode) says:

> Decode will find the next PEM formatted block (certificate, private key etc) in the input. It returns that block and the remainder of the input.

Which means the chain block might have multiple certificates. We probably need a similar decoder here.

Reply:
Sorry, cb69b8a.

Reply:
That's true. I should rephrase that. It's grabbing the first two certificates and ignoring the rest, whereas ours is loading them exhaustively into `chain_pems`.

Not sure that's a problem, but it's doing something different. The Java client does what ours does though.

Reply:
Looks good.